File: cc/050.cc
// Project Euler Problem 050
//
// Which prime, below one-million, can be written as the sum of the
// most consecutive primes?

#include <cstdio>
#include "primes1.h"


int main(int argc, const char* argv[]) {
  int longest = 21;  // Given in problem
  longest = 5;
  Primes1<int> primes;
  vector<int> sums;
  sums.push_back(0);
  int i = 0;
  while ((i<longest) || (sums[i] - sums[i-longest+1] < 1e6)) {
    sums.push_back(sums[i] + primes.next());
    i++;
  }
  int sum = 0;
  for (int j=0; j<=i-longest; j++) {
    for (int k=j+longest; k<i; k++) {
      int s = sums[k] - sums[j];
      if (s>1e6) break;
      if (!primes.is_prime(s)) continue;
      if ((k-j) < longest) continue;
      sum = s;
      longest = k-j;
    }
  }
  printf("%d\n", sum);
}

File: go/problems/050.go
// Project Euler Problem 050
//
// Which prime, below one-million, can be written as the sum of the
// most consecutive primes?

package problems

import (
	"fmt"

	"github.com/zellyn/project-euler/go/primes"
)

func Problem050() string {
	sums := []int{}
	sum := 0
	for n := 0; ; n++ {
		if p := primes.Get(n); p < 1000000 {
			sum += p
			sums = append(sums, sum)
		} else {
			break
		}
	}
	longest := 0
	prime := 0
	for i := 0; i < len(sums)-1; i++ {
		for j := i + 1 + longest; j < len(sums); j++ {
			sum := sums[j] - sums[i]
			if sum >= 1000000 {
				break
			}
			if primes.IsPrime(sum) {
				longest = j - i
				prime = sum
			}
		}
	}
	return fmt.Sprint(prime)
}

func init() {
	Functions[50] = Problem050
}
